### ISSUE TITLE
startJVM throws EXCEPTION_ACCESS_VIOLATION

### ISSUE THREAD

**Issue by ap--** [jpype-project/jpype#1023]
Hello,

I am currently investigating an error on windows that came up for some of our users https://github.com/bayer-science-for-a-better-life/paquo/issues/67 And I was wondering if this a known issue for jpype.

Basically startJVM throws the following error and crashes the python interpreter:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  EXCEPTION_ACCESS_VIOLATION (0xc0000005) at pc=0x0000000000000000, pid=9324, tid=14620
#
# JRE version:  (16.0.2+7) (build )
# Java VM: OpenJDK 64-Bit Server VM (16.0.2+7, mixed mode, tiered, compressed oops, compressed class ptrs, g1 gc, windows-amd64)
# Problematic frame:
# C  0x0000000000000000
#
# No core dump will be written. Minidumps are not enabled by default on client versions of Windows
#
# An error report file with more information is saved as:
# C:\Users\mail\Development\paquo\hs_err_pid9324.log
```

Here's a minimal script to reproduce the error:
(most of it is just to download the tool which we wrap using jpype together with its jvm)

```python
import os.path
import platform

assert platform.system() == "Windows", "this code is meant for windows"

def download_qupath():
    from urllib.request import urlopen
    from urllib.parse import urlsplit
    from shutil import unpack_archive

    qp_url = "https://github.com/qupath/qupath/releases/download/v0.3.1/QuPath-0.3.1-Windows.zip"
    chunk_size = 10 * 1024 * 1024
    fn = os.path.basename(urlsplit(qp_url).path)
    with open(fn, mode="wb+") as tmp:
        with urlopen(qp_url) as f:
            print("downloading", qp_url)
            for chunk in iter(lambda: f.read(chunk_size), b""):
                print(".", end="", flush=True)
                tmp.write(chunk)
            print("OK")
    print("unpacking")
    unpack_archive(fn)


QP_DIR = "QuPath-0.3.1"
if not os.path.isdir(QP_DIR):
    download_qupath()
    print("setup done")

# Code Producing the Error
# ========================

import jpype

jvm_path = os.path.join(os.path.abspath(QP_DIR), "runtime", "bin", "server", "jvm.dll")

try:
    jpype.startJVM(jvm_path)  # crashes with EXCEPTION_ACCESS_VIOLATION
finally:
    print("... never reached ...")
```

I have tested this locally with python3.9:
```
(vevn39) C:\Users\mail\Development\paquo>python --version --version
Python 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)]
```
But users have reported the same issue on windows with python3.8

Any insight would be very much appreciated.

Cheers,
Andreas 😃

**Comment by john-hen** [comment]
This sounds very familiar to me, from this issue in my repo: MPh-py/MPh#49.

The `jvm.dll` depends on a number of external DLLs. If those aren't unique on the system, it may end up loading the wrong one. In your case, it seems to outright crash. I suspect the problem occurs when it tries to load either `vcruntime140.dll` or `vcruntime140_1.dll`, which ship with that vendored Java run-time of yours, but are commonly found elsewhere on a Windows system as well.

It helps to put the `bin` folder that they're in first on the search path:
```python
import jpype
import os

QP_DIR = "QuPath-0.3.1"

bin = os.path.join(os.path.abspath(QP_DIR), "runtime", "bin")
os.environ["PATH"] = bin + os.pathsep + os.environ["PATH"]

jvm = os.path.join(bin, "server", "jvm.dll")
jpype.startJVM(jvm)
```

**Comment by ap--** [comment]
Thank you so much @john-hen

Interestingly your suggested fix solves the issue for my python3.9 installation from python.org,
but it does not work for my python3.9 installation from the Microsoft Store. I'll try to investigate a bit further.

It's definitely a bit confusing:

#### microsoft store python3.9
```cmd
C:\Users\mail\Development\paquo>python --version --version
Python 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)]

C:\Users\mail\Development\paquo>python -c "import sys; print(sys.executable)"
C:\Users\mail\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\python.exe
```

#### python.org python3.9
```cmd
C:\Users\mail\Development\paquo>C:\Python39\python --version --version
Python 3.9.9 (tags/v3.9.9:ccb0e6a, Nov 15 2021, 18:08:50) [MSC v.1929 64 bit (AMD64)]

C:\Users\mail\Development\paquo>C:\Python39\python -c "import sys; print(sys.executable)"
C:\Python39\python.exe
```

**Comment by john-hen** [comment]
Yeah, that *is* confusing. I never use the Microsoft Store installation, only the python.org one. And I'd guess that's true for most of my Windows users as well.

But as Python 3.9 ships with `vcruntime140.dll` and `vcruntime140_1.dll`, you may want to check if those binaries are somehow different between the two distributions. There's a chance that if Python loads those libraries, then the Java VM, running in the same process, will just keep using whichever one was already loaded, so the search-path manipulation will have no effect (as far as these two DLLs are concerned.) And maybe (pure speculation) one version is incompatible with the Java RE build, the other one isn't.

As you can see from the issue I linked, not all Java RE builds depend on these specific DLLs. So if it's up to you to vendor in a different one, without that dependency, that might solve your problem. (I'm really no expert on this, so, again, pure speculation.)

**Comment by ap--** [comment]
So to summarize my findings:

@john-hen's suggested fix worked for all python.org Python installations on Windows 10.
It did not work for MicrosoftStore installed Pythons.

I couldn't work out the specific difference between the Microsoft Store Python versions (I tested py37, py38 and py39) and the python.org Pythons without investing more time, so I opted for throwing a warning in case a user is running a MicrosoftStore Python. Here's the warning code:

https://github.com/bayer-science-for-a-better-life/paquo/blob/d8314c86da16d657f0f457cb93f902fa7d2dbdcd/paquo/jpype_backend.py#L187-L198

Moreover, I tried adding the `runtime/bin` as a jvm option via `-Djava.library.path=...` but that did not work.

To minimize the changes to the runtime environment, I am currently using a context manager around `jpype.startJVM` to only change the `"PATH"` environment variable for starting the jvm, and that works for my usecase.

Thanks again for the help 😃 

Cheers,
Andreas 😃
